src/gui/DatabaseWidget.cpp

```diff
@@ -1527,21 +1529,20 @@ bool DatabaseWidget::save(int attempt)
     }
 
     blockAutoReload(true);
+    ++m_saveAttempts;
+
     // TODO: Make this async, but lock out the database widget to prevent re-entrance
     bool useAtomicSaves = config()->get("UseAtomicSaves", true).toBool();
     QString errorMessage;
     bool ok = m_db->save(&errorMessage, useAtomicSaves, config()->get("BackupBeforeSave").toBool());
     blockAutoReload(false);
 
     if (ok) {
+        m_saveAttempts = 0;
         return true;
     }
 
-    if (attempt >= 0 && attempt <= 2) {
-        return save(attempt + 1);
-    }
-
-    if (attempt > 2 && useAtomicSaves) {
+    if (m_saveAttempts > 2 && useAtomicSaves) {
```

Comment:
You are missing the part where you are actually retrying. This code is never executed unless the user retries himself. Is that intentional? I guess it would make sense this way. I would prefer the logic reversed, though. Show error and return false if saving failed and m_saveAttempts is less than 3. Otherwise continue with whatever recovery logic we have.

Reply:
Yes that is the new intention. The old method automatically tried saving 3 times in a row using recursive calls.

Reply:
The way the logic is programmed now is optimal. You want to intervene before getting to the final error message to try and disable the save safes after the third attempt. The block is never shown if atomic saves are already disabled. Reversing the logic will require code duplication or more complicated if/else blocks.

Reply:
Alright. I'm fine with that.